### PR TITLE
Generate output docs

### PR DESCRIPTION
**Pull request by aakarshg**
The main purpose of this commit is to be able to generate multiple
output docs from a single instance, as well as moving away from the
current implementation of making an instance an iterable.

In the process of the above, fixed a few bugs around exit as well
as added a new run_scribe.py, a simple script showing how to use
scribe as a python library

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,9 +33,8 @@ takes 2 inputs and can be run as follows.
 ### Using scribe as a python library
 
 If you're just looking for a way to use scribe to generate documents and
-consume them to build your JSON, which would be scribe's use case. You can do
-something like the following.
-The smallest sample would look like:
+consume them to build your JSON, which would be scribe's use case.
+You can take a look at run_scribe.py which look:
```
Comment (portante) [inline README.md]: Not sure about these two sentences.  Are you trying to say:

```
If you're just looking for a way to use `scribe` to generate documents and consume them to build your JSON, take a look at `run_scribe.py` (excerpt below):

    from transcribe.render import transcribe
    for scribe_object in transcribe('/tmp/stockpile.json', 'stockpile'):
        print(scribe_object)
```

File: transcribe/render.py [hunk 2]
```diff
@@ -29,15 +29,22 @@ def transcribe(input_path, scribe_type):
...
-                                              schema)
                     except ValueError as e:
                         print(e)
                         continue
-                    yield(json.dumps(output_doc, indent=4))
+                    except TypeError as e:
+                        print(e)
+                        continue
+                    schema = load_schema_file(key)
+                    for document in scribe_module_instace.parse():
```
Comment (portante) [inline transcribe/render.py]: s/scribe_module_instace/scribe_module_instance/

File: transcribe/render.py [hunk 3]
```diff
@@ -50,12 +57,19 @@ def transcribe(input_path, scribe_type):
...
+                schema = load_schema_file(key)
+                for document in scribe_module_instace.parse():
+                    try:
+                        output_doc = \
+                            validate_document(document,
+                                              schema)
+                    except ValueError as e:
+                        print(e)
+                        continue
+                    yield(json.dumps(output_doc, indent=4))
```
Comment (portante) [inline transcribe/render.py]: This code looks just like the same stanza above.  Can you consider yielding the raw `scribe_module_instance` (spelling fix) document, wrapped by another generator that handles the `validate_document()` call?

File: transcribe/scribe_modules/base_scribe_module.py [hunk 4]
```diff
@@ -16,15 +16,12 @@ def __init__(self, input_dict=None, module_name=None, input_type=None,
         self.source_type = input_type
         self.host = host_name
         self.scribe_uuid = scribe_uuid
-        # self.value = list(input_dict.values())[0]
+        self._dict = dict()
+        self._dict['module'] = self.module
+        self._dict['host'] = self.host
+        self._dict['source_type'] = self.source_type
+        self._dict['scribe_uuid'] = self.scribe_uuid
```
Comment (portante) [inline transcribe/scribe_modules/base_scribe_module.py]: Nit, rewrite as:

```
self._dict = dict(module=self.module, host=self.host,
        source_type=self.source_type, scribe_uuid=self.scribe_uuid)
```

File: transcribe/scribe_modules/cpu_vulnerabilities.py [hunk 5]
```diff
@@ -12,19 +14,13 @@ def __init__(self, input_dict=None, module_name=None, host_name=None,
...
-            setattr(self, key, initial_data[key])
+            self._input_dict = input_dict
+        else:
+            raise ValueError('Input dictionary is empty')
 
-    def _parse(self, input_parse):
-        output_dict = {}
+    def parse(self):
+        input_parse = self._input_dict
+        output_dict = copy.deepcopy(self._dict)
```
Comment (portante) [inline transcribe/scribe_modules/cpu_vulnerabilities.py]: Deep copy can be very, very expensive.  The only time you need a copy is if you are going to expose a python object that somebody is going to modify.  Can we avoid this by fixing the need to modify the objects exposed?

File: transcribe/scribe_modules/cpuinfo.py [hunk 6]
```diff
@@ -56,4 +56,6 @@ def _parse(self, cpu_fullinput):
             # We update our dictionary with the created Flags list above
             current_dict.update({'Flags': new_flags})
             output_data[split_lines[0]].append(current_dict)
-        return output_data
+        output_dict = copy.deepcopy(self._dict)
```
Comment (portante) [inline transcribe/scribe_modules/cpuinfo.py]: Again `deepcopy` is expensive.  We need to consider this well.

File: transcribe/scribe_modules/yum_repos.py [hunk 7]
```diff
@@ -16,19 +18,17 @@ def __init__(self, input_dict=None, module_name=None, host_name=None,
...
-            # This is just for the sake of it
-            self.base_url = format_url(base_url, self.repo_name)
-
-    def update_repo_state(self, value):
-        if value['state'] == 'enabled':
-            return 1
-        return 0
+            self._input_dict = input_dict
+        else:
+            raise ValueError('Input dictionary is empty')
```
Comment (portante) [inline transcribe/scribe_modules/yum_repos.py]: This pattern is repeated so many times across all the modules.  Seems like it is worth to follow the DRY principle here and find a way to abstract that away it is done the same way, once, and done correctly.

**Review by portante — CHANGES_REQUESTED** [review]
Good start!

File: transcribe/scribe_modules/base_scribe_module.py (hunk 4 above)
Comment (aakarshg) [reply transcribe/scribe_modules/base_scribe_module.py]: nice catch

File: transcribe/scribe_modules/cpu_vulnerabilities.py (hunk 5 above)
Comment (aakarshg) [reply transcribe/scribe_modules/cpu_vulnerabilities.py]: We want to yield multiple objects per invocation, so i'm not sure there'll be a way to do that.

File: transcribe/scribe_modules/yum_repos.py (hunk 7 above)
Comment (aakarshg) [reply transcribe/scribe_modules/yum_repos.py]: Agreed, i'll just create a simple function in a lib and use that

File: transcribe/scribe_modules/cpu_vulnerabilities.py (hunk 5 above)
Comment (aakarshg) [reply transcribe/scribe_modules/cpu_vulnerabilities.py]: nvm its not needed in a few modules.. i am going to remove that

File: transcribe/scribe_modules/base_scribe_module.py [hunk 8]
```diff
@@ -16,15 +16,14 @@ def __init__(self, input_dict=None, module_name=None, input_type=None,
         self.source_type = input_type
         self.host = host_name
         self.scribe_uuid = scribe_uuid
-        # self.value = list(input_dict.values())[0]
+        self._dict = dict(module=self.module, host=self.host,
+                          source_type=self.source_type,
+                          scribe_uuid=self.scribe_uuid))
```
Comment (rsevilla87) [inline transcribe/scribe_modules/base_scribe_module.py]: Typo here. Remove trailing )

File: transcribe/scribe_modules/cpu_vulnerabilities.py [hunk 9]
```diff
@@ -1,6 +1,8 @@
 from . import ScribeModuleBaseClass
 from . lib.util import format_url
 
+import copy
```
Comment (rsevilla87) [inline transcribe/scribe_modules/cpu_vulnerabilities.py]: No longer required

File: transcribe/scribe_modules/cpuinfo.py [hunk 10]
```diff
@@ -1,9 +1,10 @@
 from . import ScribeModuleBaseClass
-from . lib.util import format_url
+from . lib.util import format_url, validate_length
 
 import re as _re
 import sys
-
+import json
+import copy
```
Comment (rsevilla87) [inline transcribe/scribe_modules/cpuinfo.py]: No longer required

File: transcribe/scribe_modules/dmidecode.py [hunk 11]
```diff
@@ -3,6 +3,7 @@
 
 import re as _re
 import sys
+import copy
```
Comment (rsevilla87) [inline transcribe/scribe_modules/dmidecode.py]: No longer required

File: transcribe/scribe_modules/k8s_namespaces.py [hunk 12]
```diff
@@ -1,9 +1,10 @@
 from . import ScribeModuleBaseClass
-from . lib.util import to_list
+from . lib.util import to_list, validate_length
 
 import re as _re
 import sys
 import json
+import copy
```
Comment (rsevilla87) [inline transcribe/scribe_modules/k8s_namespaces.py]: No longer required

File: transcribe/scribe_modules/k8s_nodes.py [hunk 13]
```diff
@@ -1,8 +1,9 @@
 from . import ScribeModuleBaseClass
-from . lib.util import to_list
+from . lib.util import to_list, validate_length
 
 import re as _re
 import sys
+import copy
```
Comment (rsevilla87) [inline transcribe/scribe_modules/k8s_nodes.py]: No longer required

File: transcribe/scribe_modules/yum_repos.py [hunk 14]
```diff
@@ -1,6 +1,8 @@
 from . import ScribeModuleBaseClass
 from . lib.util import format_url
 
+import copy
```
Comment (rsevilla87) [inline transcribe/scribe_modules/yum_repos.py]: No longer required

File: transcribe/scribe_modules/yum_repos.py [hunk 15]
```diff
@@ -15,20 +17,14 @@ def __init__(self, input_dict=None, module_name=None, host_name=None,
...
-        return 0
 
-    def __iter__(self):
-        for attr, value in self.__dict__.items():
-            yield attr, value
+    def parse(self):
+        def update_repo_state(self, value):
+            if value['state'] == 'enabled':
+                return 1
+            return 0
```
Comment (rsevilla87) [inline transcribe/scribe_modules/yum_repos.py]: Suggestion, this nested method could be easily written like:

```output_dict['repo_state'] = 1 if self._input_dict["value"] == "enabled" else 0```

**Review by rsevilla87 — COMMENTED** [review]
@aakarshg I've added some other comments and suggestions.

**Review by jtaleric — APPROVED** [review]
LGTM, nice work @aakarshg

**Comment by rsevilla87** [comment]
:+1:  @aakarshg, LGTM

**Review by rsevilla87 — APPROVED** [review]

File: transcribe/render.py [hunk 16]
```diff
@@ -50,12 +57,19 @@ def transcribe(input_path, scribe_type):
...
-                        validate_document(dict(scribe_module_instace),
-                                          schema)
                 except ValueError as e:
                     print(e)
                     continue
-                yield(json.dumps(output_doc, indent=4))
+                except TypeError as e:
+                    print(e)
+                    continue
+                schema = load_schema_file(key)
```
Comment (portante) [inline transcribe/render.py]: These lines, 66 - 75, look identical to lines 38 - 47.  Any way to apply the DRY principle here?

**Review by portante — CHANGES_REQUESTED** [review]
Looks pretty good, just DRY ... ;)

File: transcribe/render.py [hunk 17]
```diff
@@ -7,55 +7,63 @@
...
+                                         host_name=host_dict,
+                                         input_type=input_type,
+                                         scribe_uuid=scribe_uuid)
+    except ImportError as e:
+        print(e)
+        return e
+    except ValueError as e:
+        print(e)
+        return e
+    except TypeError as e:
```
Comment (portante) [inline transcribe/render.py]: These three statements can be collapsed to:

```
except (ImportError, ValueError, TypeError) as e:
    print(e)
    return e
```

File: transcribe/render.py [hunk 18]
```diff
@@ -7,55 +7,63 @@
...
-                                              schema)
-                    except ValueError as e:
-                        print(e)
-                        continue
-                    yield(json.dumps(output_doc, indent=4))
+                    yield from _create_module_yield_doc(key, value,
+                                                                 host_dict['host'],
+                                                                 _input_type,
+                                                                 _scribe_uuid)
+                    # for document in _create_module_yield_doc(key, value,
```
Comment (portante) [inline transcribe/render.py]: Do you need to keep this commented out?

File: transcribe/render.py [hunk 19]
```diff
@@ -7,55 +7,63 @@
...
-                except ValueError as e:
-                    print(e)
-                    continue
-                yield(json.dumps(output_doc, indent=4))
+                yield from _create_module_yield_doc(key,
+                                                         host_dict['value'],
+                                                         host_dict['host'],
+                                                         _input_type,
+                                                         _scribe_uuid)
+                # for document in _create_module_yield_doc(key,
```
Comment (portante) [inline transcribe/render.py]: Same here?

File: transcribe/render.py [hunk 20]
```diff
@@ -7,55 +7,63 @@
...
-                try:
-                    output_doc = \
-                        validate_document(dict(scribe_module_instace),
-                                          schema)
-                except ValueError as e:
-                    print(e)
-                    continue
-                yield(json.dumps(output_doc, indent=4))
+                yield from _create_module_yield_doc(key,
+                                                         host_dict['value'],
```
Comment (portante) [inline transcribe/render.py]: Seems like inconsistent formatting here ...

File: transcribe/render.py [hunk 21]
```diff
@@ -7,55 +7,63 @@
...
-                    try:
-                        output_doc = \
-                            validate_document(dict(scribe_module_instace),
-                                              schema)
-                    except ValueError as e:
-                        print(e)
-                        continue
-                    yield(json.dumps(output_doc, indent=4))
+                    yield from _create_module_yield_doc(key, value,
+                                                                 host_dict['host'],
```
Comment (portante) [inline transcribe/render.py]: Seems like inconsistent formatting here ...

**Review by portante — CHANGES_REQUESTED** [review]
A few more considerations ...

File: transcribe/render.py (hunk 18 above)
Comment (aakarshg) [reply transcribe/render.py]: oopsie sent this in for ci to vet sorry

**Review by portante — APPROVED** [review]
Looks good!

**Pull request merged**